File: luatests/lua/ipairs.lua
local function ipairs_iterator(t, n)
    if n < #t then
        return n + 1, t[n + 1]
    end
end

local function _ipairs(t)
    return ipairs_iterator, t, 0
end

local t = {5, 4, 3}
local s = 0

for i, v in _ipairs(t) do
    s = s + i * v
end
print(s) -- 5*1 + 4*2 + 3*3
--> =22

File: lib/base/ipairs.go
package base

import rt "github.com/arnodel/golua/runtime"

func ipairsIteratorF(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	coll := c.Arg(0)
	n, err := c.IntArg(1)
	if err != nil {
		return nil, err
	}
	next := c.Next()
	n++
	nv := rt.IntValue(n)
	v, err := rt.Index(t, coll, nv)
	if err != nil {
		return nil, err
	}
	if !v.IsNil() {
		t.Push1(next, nv)
		t.Push1(next, v)
	}
	return next, nil
}

var ipairsIterator = rt.NewGoFunction(ipairsIteratorF, "ipairsiterator", 2, false)

func ipairs(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	next := c.Next()
	t.Push1(next, rt.FunctionValue(ipairsIterator))
	t.Push1(next, c.Arg(0))
	t.Push1(next, rt.IntValue(0))
	return next, nil
}
